clippy_lints/src/zero_repeat_side_effects.rs

```diff
@@ -45,20 +46,44 @@ declare_clippy_lint! {
 declare_lint_pass!(ZeroRepeatSideEffects => [ZERO_REPEAT_SIDE_EFFECTS]);
 
 impl LateLintPass<'_> for ZeroRepeatSideEffects {
-    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &'_ rustc_hir::Expr<'_>) {
+    fn check_expr(&mut self, cx: &LateContext<'_>, expr: &rustc_hir::Expr<'_>) {
+        let hir_map = cx.tcx.hir();
         if let Some(args) = VecArgs::hir(cx, expr)
             && let VecArgs::Repeat(inner_expr, len) = args
             && let ExprKind::Lit(l) = len.kind
             && let LitKind::Int(i, _) = l.node
             && i.0 == 0
         {
             inner_check(cx, expr, inner_expr, true);
-        } else if let ExprKind::Repeat(inner_expr, _) = expr.kind
-            && let ty::Array(_, cst) = cx.typeck_results().expr_ty(expr).kind()
-            && let ConstKind::Value(_, ty::ValTree::Leaf(element_count)) = cst.kind()
-            && element_count.to_target_usize(cx.tcx) == 0
-        {
-            inner_check(cx, expr, inner_expr, false);
+        } else {
+            let ExprKind::Repeat(inner_expr, length) = expr.kind else {
+                return;
+            };
+            // Skip if the length is from a macro.
+            if let ArrayLen::Body(anon_const) = length {
+                let length_expr = hir_map.body(anon_const.body).value;
+                if !length_expr.span.eq_ctxt(expr.span) {
+                    return;
+                }
+
+                // Get the initialization span of a const item and compare it with the span at use-site.
+                if let ExprKind::Path(QPath::Resolved(None, path)) = length_expr.kind
+                    && let Res::Def(DefKind::Const, def_id) = path.res
+                    && let Some(def_id) = def_id.as_local()
+                    && let Node::Item(item) = cx.tcx.hir_node_by_def_id(def_id)
+                    && let ItemKind::Const(_ty, _, body_id) = item.kind
+                    && let init_span = hir_map.span_with_body(body_id.hir_id)
+                    && !init_span.eq_ctxt(length_expr.span)
```

Comment:
FWIW, with this change we also no longer warn something like:
```rs
const LEN: usize = 0;

macro_rules! m {
  () => {
    [drop(()); LEN]
  }
}
m!();
```
Even though the const item is not from a macro, the syntax contexts still differ. Maybe we could have another `init_span.from_expansion()` to continue warning that?

(I'm personally still of the opinion that we shouldn't lint paths to consts at all and treat them opaquely and only lint on `[expr; 0]` specifically like how the lint already does it for `vec!`, because when I read `[f(); LEN]` I intuitively expect that the function is called and it doesn't seem as confusing as `[f(); 0]`. It would also have false positives when eg. the const item depends on `#[cfg]`s and has different values in different compilation sessions)

Reply:
Yeah, I agree with your reasonings.